src/Control/Algebra.hs

```diff
@@ -79,51 +82,55 @@ type Has eff sig m = (Members eff sig, Algebra sig m)
 
 -- | Construct a request for an effect to be interpreted by some handler later on.
 send :: (Member eff sig, Algebra sig m) => eff m a -> m a
-send = alg . inj
+send = alg id . inj
 {-# INLINE send #-}
 
 
 -- base
 
 instance Algebra (Lift IO) IO where
-  alg (LiftWith with k) = with (Identity ()) coerce >>= k . runIdentity
+  alg hom (LiftWith with k) = with (Identity ()) (coerce . hom . runIdentity) >>= hom . k . runIdentity
 
 instance Algebra (Lift Identity) Identity where
-  alg (LiftWith with k) = with (Identity ()) coerce >>= k . runIdentity
+  alg hom (LiftWith with k) = with (Identity ()) (coerce . hom . runIdentity) >>= hom . k . runIdentity
 
 instance Algebra Choose NonEmpty where
-  alg (Choose m) = m True S.<> m False
+  alg hom (Choose m) = hom (m True) S.<> hom (m False)
 
 instance Algebra Empty Maybe where
-  alg Empty = Nothing
+  alg _ Empty = Nothing
 
 instance Algebra (Error e) (Either e) where
-  alg (L (Throw e))     = Left e
-  alg (R (Catch m h k)) = either (k <=< h) k m
+  alg hom = \case
+    L (Throw e)     -> Left e
+    R (Catch m h k) -> either (hom . k <=< hom . h) (hom . k) (hom m)
 
 instance Algebra (Reader r) ((->) r) where
-  alg (Ask       k) r = k r r
-  alg (Local f m k) r = k (m (f r)) r
+  alg hom = \case
+    Ask       k -> \ r -> hom (k r) r
+    Local f m k -> \ r -> hom (k (hom m (f r))) r
 
 instance Algebra NonDet [] where
-  alg (L Empty)      = []
-  alg (R (Choose k)) = k True ++ k False
+  alg hom = \case
+    L Empty      -> []
+    R (Choose k) -> hom (k True) ++ hom (k False)
 
 instance Monoid w => Algebra (Writer w) ((,) w) where
-  alg (Tell w (w', k))    = (mappend w w', k)
-  alg (Listen (w, a) k)   = let (w', a') = k w a in (mappend w w', a')
-  alg (Censor f (w, a) k) = let (w', a') = k a in (mappend (f w) w', a')
+  alg hom = \case
+    Tell w     k -> let (w', k') = hom k in (mappend w w', k')
+    Listen m   k -> let (w, a) = hom m ; (w', a') = hom (k w a) in (mappend w w', a')
+    Censor f m k -> let (w, a) = hom m ; (w', a') = hom (k a) in (mappend (f w) w', a')
 
 
 -- transformers
 
 instance (Algebra sig m, Effect sig) => Algebra (Error e :+: sig) (Except.ExceptT e m) where
-  alg (L (L (Throw e)))     = Except.throwE e
-  alg (L (R (Catch m h k))) = Except.catchE m h >>= k
-  alg (R other)             = Except.ExceptT $ alg (thread (Right ()) (either (pure . Left) Except.runExceptT) other)
+  alg hom = \case
+    L (L (Throw e))     -> Except.throwE e
+    L (R (Catch m h k)) -> Except.catchE (hom m) (hom . h) >>= hom . k
+    R other             -> Except.ExceptT $ alg id (thread (Right ()) (either (pure . Left) (Except.runExceptT . hom)) other)
```

Comment:
Nonterminal algebras, on the other hand, also need to apply the homomorphism before threading through the rest of the signature.